File: src/PDFKeeper.Presentation/Presenters/AddPdfPresenter.vb
'******************************************************************************
'* PDFKeeper -- Open Source PDF Document Management
'* Copyright (C) 2009-2022 Robert F. Frasca
'*
'* This file is part of PDFKeeper.
'*
'* PDFKeeper is free software: you can redistribute it and/or modify
'* it under the terms of the GNU General Public License as published by
'* the Free Software Foundation, either version 3 of the License, or
'* (at your option) any later version.
'*
'* PDFKeeper is distributed in the hope that it will be useful,
'* but WITHOUT ANY WARRANTY; without even the implied warranty of
'* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'* GNU General Public License for more details.
'*
'* You should have received a copy of the GNU General Public License
'* along with PDFKeeper.  If not, see <http://www.gnu.org/licenses/>.
'******************************************************************************
Imports PDFKeeper.Common
Imports PDFKeeper.Domain
Imports PDFKeeper.FileIO
Imports PDFKeeper.Infrastructure
Imports PDFKeeper.Services

Public Class AddPdfPresenter
    Private ReadOnly view As IAddPdfView
    Private ReadOnly authorListSvc As IAuthorListService
    Private ReadOnly subjectListSvc As ISubjectListService
    Private ReadOnly categoryListSvc As ICategoryListService
    Private ReadOnly taxYearListSvc As ITaxYearListService
    Private ReadOnly pdfSvc As IPdfService
    Private ReadOnly commonDialogs As New CommonDialogs
    Private viewInstance As Form
    Private password As SecureString
    Private modifiedPdfFile As String
    Private bypassClosingPrompt As Boolean

    Public Sub New(ByVal view As IAddPdfView, ByVal authorListSvc As IAuthorListService,
                   ByVal subjectListSvc As ISubjectListService, ByVal categoryListSvc As ICategoryListService,
                   ByVal taxYearListSvc As ITaxYearListService, ByVal pdfSvc As IPdfService)
        Me.view = view
        Me.authorListSvc = authorListSvc
        Me.subjectListSvc = subjectListSvc
        Me.categoryListSvc = categoryListSvc
        Me.taxYearListSvc = taxYearListSvc
        Me.pdfSvc = pdfSvc
    End Sub

    Friend Sub AddPdfDialog_Load(sender As Object, e As EventArgs)
        viewInstance = CType(sender, Form)
        With view
            .SaveEnabled = False
            .PreviewEnabled = False
            .AddEnabled = False
            .ShowOpenPdfFileDialog()
            If .SelectedPdf.Length > 0 Then
                bypassClosingPrompt = True
                Dim pdfPasswordType = pdfSvc.GetPdfPasswordType(.SelectedPdf)
                If pdfPasswordType = PdfPasswordTypes.PdfPasswordType.None Then
                    UpdateView(.SelectedPdf, Nothing)
                ElseIf pdfPasswordType = PdfPasswordTypes.PdfPasswordType.Owner Then
                    Using PdfOwnerPasswordDialog
                        PdfOwnerPasswordDialog.ShowDialog()
                        If PdfOwnerPasswordDialog.DialogResult = DialogResult.OK Then
                            password = PdfOwnerPasswordDialog.Password
                        End If
                    End Using
                    If password IsNot Nothing Then
                        If password.Length > 0 Then
                            If pdfSvc.IsPdfOwnerPasswordValid(.SelectedPdf, password) = False Then
                                commonDialogs.ShowMessageBox(My.Resources.PdfOwnerPasswordInvalid, True)
                                viewInstance.Close()
                            Else
                                password.MakeReadOnly()
                                UpdateView(.SelectedPdf, password)
                            End If
                        Else
                            commonDialogs.ShowMessageBox(My.Resources.PdfOwnerPasswordRequired, True)
                            viewInstance.Close()
                        End If
                    Else
                        viewInstance.Close()
                    End If
                ElseIf pdfPasswordType = PdfPasswordTypes.PdfPasswordType.User Then
                    commonDialogs.ShowMessageBox(My.Resources.PdfContainsUserPassword, True)
                    viewInstance.Close()
                ElseIf pdfPasswordType = PdfPasswordTypes.PdfPasswordType.Unknown Then
                    commonDialogs.ShowMessageBox(My.Resources.PdfInvalid, True)
                    viewInstance.Close()
                End If
                bypassClosingPrompt = False
            Else
                viewInstance.Close()
            End If
        End With
    End Sub

    Friend Sub ViewButton_Click(sender As Object, e As EventArgs)
        pdfSvc.ShowPdfWithRestrictedViewer(view.SelectedPdf)
    End Sub

    Friend Sub SetTitleToFileNameButton_Click(sender As Object, e As EventArgs)
        view.Title = Path.GetFileNameWithoutExtension(view.SelectedPdf)
    End Sub

    Friend Sub CommonComboBox_Enter(sender As Object, e As EventArgs)
        Dim control = CType(sender, Control)
        Try
            viewInstance.Cursor = Cursors.WaitCursor
            Dim currentItem = control.Text
            If control.Name.StartsWith("Author") Then
                view.AuthorItems = authorListSvc.ListAuthors
            ElseIf control.Name.StartsWith("Subject") Then
                view.SubjectItems = subjectListSvc.ListSubjects(view.Author)
            ElseIf control.Name.StartsWith("Category") Then
                view.CategoryItems = categoryListSvc.ListCategories
            ElseIf control.Name.StartsWith("TaxYear") Then
                view.TaxYearItems = taxYearListSvc.ListRangeOfTaxYears
            End If
            control.Text = currentItem
        Catch ex As DbException
            commonDialogs.ShowMessageBox(ex.Message, True)
        Finally
            viewInstance.Cursor = Cursors.Default
        End Try
    End Sub

    Friend Sub Common_TextChanged(sender As Object, e As EventArgs)
        With view
            If .Title.Trim.Length > 0 And .Author.Trim.Length > 0 And .Subject.Trim.Length > 0 Then
                .SaveEnabled = True
            Else
                .SaveEnabled = False
            End If
            .PreviewEnabled = False
            If .AddEnabled Then
                .AddEnabled = False
                pdfSvc.CloseRestrictedViewer()
            End If
        End With
    End Sub

    Friend Sub SaveButton_Click(sender As Object, e As EventArgs)
        Dim control = CType(sender, Control)
        With view
            Dim model = New PdfInfoModel
            model.Title = .Title
            model.Author = .Author
            model.Subject = .Subject
            model.Keywords = .Keywords
            pdfSvc.CloseRestrictedViewer()
            pdfSvc.WritePdfWithInfo(.SelectedPdf, password, modifiedPdfFile, model)
            .SaveEnabled = False
            .PreviewEnabled = True
            .AddEnabled = True
        End With
        viewInstance.SelectNextControl(control, True, True, True, True)
    End Sub

    Friend Sub PreviewButton_Click(sender As Object, e As EventArgs)
        pdfSvc.ShowPdfWithRestrictedViewer(modifiedPdfFile)
    End Sub

    Friend Sub AddButton_Click(sender As Object, e As EventArgs)
        bypassClosingPrompt = True
        Dim model = New PdfInfoExtModel
        With view
            model.Category = .Category
            model.TaxYear = .TaxYear
            If .FlagDocument Then
                model.Flag = 1
            Else
                model.Flag = 0
            End If
            pdfSvc.CloseRestrictedViewer()
            pdfSvc.WritePdfInfoExt(modifiedPdfFile, model)
            pdfSvc.StagePdfForUpload(modifiedPdfFile)
            If My.Settings.AddPdfDeleteSelectedPdfWhenAdded Then
                Dim file = New FileInfo(.SelectedPdf)
                file.DeleteToRecycleBin
            End If
        End With
        viewInstance.Close()
    End Sub

    Friend Sub AddPdfDialog_FormClosing(sender As Object, e As FormClosingEventArgs)
        If view.SelectedPdf.Length > 0 And bypassClosingPrompt = False Then
            If commonDialogs.ShowQuestionMessageBox(My.Resources.CancelQuestion, False) = DialogResult.Yes Then
                pdfSvc.CloseRestrictedViewer()
                If modifiedPdfFile IsNot Nothing Then
                    IO.File.Delete(modifiedPdfFile)
                End If
            Else
                e.Cancel = True
            End If
        End If
    End Sub

    Private Sub UpdateView(ByVal pdfFile As String, ByVal password As SecureString)
        modifiedPdfFile = Path.Combine(AppFolders.GetPath(AppFolders.AppFolder.Temp),
                                       Path.GetFileName(pdfFile))
        Dim model = pdfSvc.ReadPdfInfo(pdfFile, password)
        With view
            .Title = model.Title
            .Author = model.Author
            .Subject = model.Subject
            .Keywords = model.Keywords
        End With
    End Sub
End Class

File: src/PDFKeeper.Core/Presenters/AddPdfPresenter.cs
// ****************************************************************************
// * PDFKeeper -- Open Source PDF Document Management
// * Copyright (C) 2009-2025 Robert F. Frasca
// *
// * This file is part of PDFKeeper.
// *
// * PDFKeeper is free software: you can redistribute it and/or modify it
// * under the terms of the GNU General Public License as published by the
// * Free Software Foundation, either version 3 of the License, or (at your
// * option) any later version.
// *
// * PDFKeeper is distributed in the hope that it will be useful, but WITHOUT
// * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
// * more details.
// *
// * You should have received a copy of the GNU General Public License along
// * with PDFKeeper. If not, see <https://www.gnu.org/licenses/>.
// ****************************************************************************

using PDFKeeper.Core.Commands;
using PDFKeeper.Core.Extensions;
using PDFKeeper.Core.DataAccess;
using PDFKeeper.Core.Properties;
using PDFKeeper.Core.ViewModels;
using System;
using System.IO;
using System.Linq;
using PDFKeeper.Core.Models;
using PDFKeeper.Core.Services;
using PDFKeeper.Core.FileIO.PDF;

namespace PDFKeeper.Core.Presenters
{
    public class AddPdfPresenter : PresenterBase<AddPdfViewModel>
    {
        private readonly IFileDialogService openFileDialogService;
        private readonly IPasswordDialogService passwordDialogService;
        private readonly IRestrictedPdfViewerService restrictedPdfViewerService;
        private readonly IMessageBoxService messageBoxService;
        private PdfFile pdfFile;
        private PdfMetadata pdfMetadata;

        /// <summary>
        /// Initializes a new instance of the AddPdfPresenter class.
        /// </summary>
        /// <param name="openFileDialogService">The OpenFileDialogService instance.</param>
        /// <param name="passwordDialogService">The PasswordDialogService instance.</param>
        /// <param name="restrictedPdfViewerService">
        /// The RestrictedPdfViewerService instance.
        /// </param>
        /// <param name="messageBoxService">The MessageBoxService instance.</param>
        public AddPdfPresenter(IFileDialogService openFileDialogService,
            IPasswordDialogService passwordDialogService,
            IRestrictedPdfViewerService restrictedPdfViewerService,
            IMessageBoxService messageBoxService)
        {
            this.openFileDialogService = openFileDialogService;
            this.passwordDialogService = passwordDialogService;
            this.restrictedPdfViewerService = restrictedPdfViewerService;
            this.messageBoxService = messageBoxService;
            ViewModel = new AddPdfViewModel();
        }

        public void GetCollections()
        {
            try
            {
                ViewModel.Authors = ColumnData.GetAuthors(null, null, null).OrderBy(
                    author => author).ToArray();
                ViewModel.Categories = ColumnData.GetCategories(null, null, null).OrderBy(
                    category => category).ToArray();
                ViewModel.TaxYears = TaxYear.GetYearRange().ToArray();
            }
            catch (DatabaseException ex)
            {
                this.messageBoxService.ShowMessage(ex.Message, true);
            }
        }

        /// <summary>
        /// Prompts the user to select the PDF from the file system.If the PDF contains an Owner
        /// password, the user will be prompted to enter it.
        /// </summary>
        public void SelectPdf()
        {
            var selectedPdfPath = openFileDialogService.ShowDialog(Resources.PdfFilter, null);
            if (selectedPdfPath.Length > 0)
            {
                try
                {
                    pdfFile = new PdfFile(new FileInfo(selectedPdfPath));
                    var passwordType = pdfFile.GetPasswordType();
                    if (passwordType.Equals(PdfFile.PasswordType.None))
                    {
                        pdfMetadata = new PdfMetadata(pdfFile, null);
                        ViewModel.SelectedPdf = pdfFile.FullName;
                        ViewModel.UploadProfile = pdfMetadata.ExportUploadProfile();
                    }
                    else if (passwordType.Equals(PdfFile.PasswordType.Owner))
                    {
                        var pdfOwnerPassword = passwordDialogService.ShowDialog();
                        if (pdfOwnerPassword != null)
                        {
                            if (pdfOwnerPassword.Length > 0)
                            {
                                try
                                {
                                    pdfMetadata = new PdfMetadata(pdfFile, pdfOwnerPassword);
                                    ViewModel.SelectedPdf = pdfFile.FullName;
                                    ViewModel.UploadProfile = pdfMetadata.ExportUploadProfile();
                                }
                                catch (ArgumentException)
                                {
                                    messageBoxService.ShowMessage(Resources.PdfOwnerPasswordIncorrect,
                                        true);
                                    OnViewCloseRequested();
                                }
                                pdfOwnerPassword.MakeReadOnly();
                            }
                            else
                            {
                                messageBoxService.ShowMessage(Resources.PdfOwnerPasswordRequired,
                                    true);
                                OnViewCloseRequested();
                            }
                        }
                        else
                        {
                            OnViewCloseRequested();
                        }
                    }
                    else if (passwordType.Equals(PdfFile.PasswordType.User))
                    {
                        messageBoxService.ShowMessage(Resources.PdfContainsUserPassword, true);
                        OnViewCloseRequested();
                    }
                    else if (passwordType.Equals(PdfFile.PasswordType.Unknown))
                    {
                        messageBoxService.ShowMessage(Resources.PdfInvalid, true);
                        OnViewCloseRequested();
                    }
                }
                catch (ArgumentException ex)
                {
                    messageBoxService.ShowMessage(ex.Message, true);
                    OnViewCloseRequested();
                }
            }
            else
            {
                OnViewCloseRequested();
            }
        }

        public void ViewPdf()
        {
            restrictedPdfViewerService.Show(ViewModel.SelectedPdf);
        }

        public void SetTitleToPdfFileName()
        {
            OnApplyPendingChangesRequested();
            ViewModel.Title = Path.GetFileNameWithoutExtension(ViewModel.SelectedPdf);
        }

        public void GetSubjects()
        {
            try
            {
                var entry = ViewModel.Subject;
                ViewModel.Subjects = ColumnData.GetSubjects(ViewModel.Author, null, null).OrderBy(
                    subject => subject).ToArray();
                ViewModel.Subject = entry;
            }
            catch (DatabaseException ex)
            {
                this.messageBoxService.ShowMessage(ex.Message, true);
            }
        }

        /// <summary>
        /// Adds a copy of the PDF with the specified information metadata applied and the
        /// corresponding XML containing the specified external metadata to the UploadStaging
        /// folder.
        /// </summary>
        /// <param name="deleteSourcePdf">
        /// To delete the source PDF to the Operating System Recycle Bin(true or false).
        /// </param>
        public void AddPdf(bool deleteSourcePdf)
        {
            CancelViewClosing = false;
            OnApplyPendingChangesRequested();
            pdfMetadata.ImportUploadProfile(ViewModel.UploadProfile);
            restrictedPdfViewerService.Close();
            try
            {
                var targetPdfFile = pdfMetadata.Write();
                new UploadStagingCommand(targetPdfFile).Execute();
                if (deleteSourcePdf)
                {
                    new FileInfo(pdfFile.FullName).DeleteToRecycleBin();
                }
            }
            catch (NullReferenceException ex)
            {
                messageBoxService.ShowMessage(ex.Message, true);
                OnViewCloseCancelled();
            }
            catch (iText.IO.Exceptions.IOException ex)
            {
                messageBoxService.ShowMessage(ex.Message, true);
                OnViewCloseCancelled();
            }
        }

        public void Cancel()
        {
            CancelViewClosing = false;
            if (messageBoxService.ShowQuestion(Resources.CancelQuestion, false).Equals(7))
            {
                OnViewCloseCancelled();
            }
            else
            {
                restrictedPdfViewerService.Close();
            }
        }
    }
}
